File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    static boolean[] visited;
    
    public int solution(int n, int[][] computers) {
        int answer = 0;
        visited = new boolean[n];
        
        for (int v = 0; v < n; v++) {
            if (!visited[v]) {
                bfs(v, computers);
                answer++;
            }
        }
        
        return answer;
    }
    
    void bfs(int start, int[][] computers) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(start);
        visited[start] = true;
        
        while(!queue.isEmpty()) {
            int curr = queue.poll();
            
            for (int i = 0; i < computers.length; i++) {
                if (computers[curr][i] == 1 && !visited[i]) {
                    queue.add(i);
                    visited[i] = true;
                }
            }
        }
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import deque

def solution(n, computers):
    answer = 0
    visited = [False for _ in range(n)]
    
    def bfs(v):
        visited[v] = True # 방문처리
        
        dq = deque()
        dq.append(v)
        
        while dq:
            curr = dq.popleft()
            
            for i in range(n):
                if visited[i] == False and computers[curr][i] == 1:
                    visited[i] = True
                    dq.append(i)
    
    for v in range(n): # 시작지점이 정해져 있지 않아 모든 노드를 다 검사해야 함
        if not visited[v]:
            bfs(v)
            answer += 1
    
    return answer